### PR TITLE
Avoid materializing COW inputs to many forward ops

### PR DESCRIPTION
**Pull request by kurtamohler**
This PR updates many of the forward operations to avoid materializing COW inputs. It also adds a test to `test_ops.py` which calls all of the operations with COW tensor inputs and checks that they do not get materialized if the OpInfo is marked as avoiding COW materialization. Also, if a COW input is not materialized, the test checks that its data does not get mutated

There are still a lot of remaining forward operations that need to be updated to avoid materializing COW inputs. These have been marked with `supports_cow_input_no_materialize=False` in `common_methods_invocations.py` and will be addressed in a later PR. Backward operations will need to be addressed in later PRs as well

Part of #97856 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118926

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/core/TensorAccessor.h [hunk 1]
```diff
@@ -18,6 +18,11 @@ struct DefaultPtrTraits {
   typedef T* PtrType;
 };
 
+template <typename T>
+struct ConstPtrTraits {
```
Comment (kurtamohler) [inline aten/src/ATen/core/TensorAccessor.h:22]: TODO: I realized that I don't actually need this or the `TensorBase::const_accessor` method. I can just give a const type `T` to `TensorBase::accessor<T,...>` and make it call `const_data_ptr` if the type is const

**Comment by kurtamohler** [comment]
If this PR is too big, I can break it up into smaller pieces

**Comment by ezyang** [comment]
How tough are splits? The main advantage to splitting is I can slowly work on it PR by PR and you can land stuff as it is ready.

File: aten/src/ATen/core/TensorBase.h [hunk 2]
```diff
@@ -620,7 +620,13 @@ class TORCH_API TensorBase {
   TensorAccessor<T,N> accessor() const& {
     static_assert(N > 0, "accessor is used for indexing tensor, for scalars use *data_ptr<T>()");
     TORCH_CHECK(dim() == N, "TensorAccessor expected ", N, " dims but tensor has ", dim());
-    return TensorAccessor<T,N>(data_ptr<T>(),sizes().data(),strides().data());
+    T* ptr = nullptr;
+    if constexpr (std::is_const<T>::value) {
+      ptr = const_data_ptr<std::remove_const_t<T>>();
```
Comment (ezyang) [inline aten/src/ATen/core/TensorBase.h:625]: IDK, should `const_data_ptr` do this removal automatically?

**Comment by kurtamohler** [comment]
> How tough are splits? The main advantage to splitting is I can slowly work on it PR by PR and you can land stuff as it is ready.

It shouldn't be too tough. I'm willing to close this PR and resubmit a stack of smaller ones

File: aten/src/ATen/core/TensorBase.h (hunk 2 above)
Comment (kurtamohler) [reply aten/src/ATen/core/TensorBase.h:625]: Yeah that seems reasonable

**Comment by kurtamohler** [comment]
Replaced with stack at #119507

**Pull request closed without merging**